rust/lanzatool/src/generation.rs

```diff
@@ -1,42 +1,69 @@
+use serde::{Deserialize, Serialize};
+use serde::de::IntoDeserializer;
+
 use std::fmt;
 use std::fs;
-use std::path::Path;
+use std::path::{Path, PathBuf};
 
-use anyhow::{Context, Result};
+use anyhow::{anyhow, Context, Result};
+use bootspec::generation::Generation as BootspecGeneration;
+use bootspec::BootJson;
+use bootspec::SpecialisationName;
 
-use crate::bootspec::Bootspec;
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SecureBootExtension {
+    #[serde(rename="osRelease")]
+    pub os_release: PathBuf
+}
 
 #[derive(Debug)]
 pub struct Generation {
+    /// Top-level nixpkgs version
     version: u64,
-    specialisation_name: Option<String>,
-    pub bootspec: Bootspec,
+    /// Top-level specialisation name
+    specialisation_name: Option<SpecialisationName>,
+    /// Top-level bootspec document
+    pub bootspec: BootJson,
+    /// SecureBoot extensions extracted from bootspec
+    pub extensions: SecureBootExtension
 }
 
 impl Generation {
     pub fn from_toplevel(toplevel: impl AsRef<Path>) -> Result<Self> {
-        let bootspec_path = toplevel.as_ref().join("bootspec/boot.v1.json");
-        let bootspec: Bootspec = serde_json::from_slice(
+        let bootspec_path = toplevel.as_ref().join("bootspec/boot.json");
+        let generation: BootspecGeneration = serde_json::from_slice(
             &fs::read(bootspec_path).context("Failed to read bootspec file")?,
         )
         .context("Failed to parse bootspec json")?;
 
+        let bootjson: BootJson = generation
+            .try_into().map_err(|err: &'static str| anyhow!(err))?;
+
+        let sb_extensions: SecureBootExtension = Deserialize::deserialize(
+            bootjson.extensions.get("lanzaboote")
```

Comment:
Is here a formatting error? Did you run `cargo fmt` on this?

Reply:
People like me don't need formatted code where they go (spoiler: hell)